app/scripts/controllers/incoming-transactions.js

```diff
@@ -1,320 +0,0 @@
- * @property {string} nonce - The transaction nonce, in decimal
- * @property {string} timeStamp - The timestamp for the transaction, in seconds
- * @property {string} to - The hex-prefixed address of the recipient
- * @property {string} value - The amount of ETH sent in this transaction, in decimal WEI
- */
-
-/**
- * This controller is responsible for retrieving incoming transactions. Etherscan is polled once every block to check
- * for new incoming transactions for the current selected account on the current network
- *
- * Note that only Etherscan-compatible networks are supported. We will not attempt to retrieve incoming transactions
- * on non-compatible custom RPC endpoints.
- */
-export default class IncomingTransactionsController {
-  constructor(opts = {}) {
-    const {
-      blockTracker,
-      onNetworkDidChange,
-      getCurrentChainId,
-      preferencesController,
-      onboardingController,
-    } = opts;
-    this.blockTracker = blockTracker;
-    this.getCurrentChainId = getCurrentChainId;
-    this.preferencesController = preferencesController;
-    this.onboardingController = onboardingController;
-
-    this._onLatestBlock = async (newBlockNumberHex) => {
-      const selectedAddress = this.preferencesController.getSelectedAddress();
-      const newBlockNumberDec = parseInt(newBlockNumberHex, 16);
-      await this._update(selectedAddress, newBlockNumberDec);
-    };
-
-    const incomingTxLastFetchedBlockByChainId = Object.keys(
-      ETHERSCAN_SUPPORTED_NETWORKS,
-    ).reduce((network, chainId) => {
-      network[chainId] = null;
-      return network;
-    }, {});
-
-    const initState = {
-      incomingTransactions: {},
-      incomingTxLastFetchedBlockByChainId,
-      ...opts.initState,
-    };
-    this.store = new ObservableStore(initState);
-
-    this.preferencesController.store.subscribe(
-      previousValueComparator((prevState, currState) => {
-        const {
-          featureFlags: {
-            showIncomingTransactions: prevShowIncomingTransactions,
-          } = {},
-        } = prevState;
-        const {
-          featureFlags: {
-            showIncomingTransactions: currShowIncomingTransactions,
-          } = {},
-        } = currState;
-
-        if (currShowIncomingTransactions === prevShowIncomingTransactions) {
-          return;
-        }
-
-        if (prevShowIncomingTransactions && !currShowIncomingTransactions) {
-          this.stop();
-          return;
-        }
-
-        this.start();
-      }, this.preferencesController.store.getState()),
-    );
-
-    this.preferencesController.store.subscribe(
-      previousValueComparator(async (prevState, currState) => {
-        const { selectedAddress: prevSelectedAddress } = prevState;
-        const { selectedAddress: currSelectedAddress } = currState;
-
-        if (currSelectedAddress === prevSelectedAddress) {
-          return;
-        }
-        await this._update(currSelectedAddress);
```

Comment:
I liked the idea of removing all these subscription methods of controllers (`preferencesController`, `onboardingController`) and using callbacks instead to read the state. But on these subscription methods, we enforce `update`s. Did we do this intentionally or a minor thing that we can avoid?

Reply:
If you mean why do we no longer do this, it's just a matter of simplifying the code.

We previously used to listen to every possible controller which could change the address and maintain a copy in our own state.

Instead we now just need a single callback to request the current address from the state whenever we need to check it.

Reply:
Great spot, have added these updates back (plus unit tests) but via listeners in `MetaMaskController` to decouple it from the helper as it's a client decision.